src/ControllerMessenger.ts

```diff
@@ -0,0 +1,89 @@
+export type ActionHandler<Action, ActionType> = (
+  ...args: ExtractActionParameters<Action, ActionType>
+) => ExtractActionResponse<Action, ActionType>;
+export type ExtractActionParameters<Action, T> = Action extends { type: T; handler: (...args: infer H) => any }
+  ? H
+  : never;
+export type ExtractActionResponse<Action, T> = Action extends { type: T; handler: (...args: any) => infer H }
+  ? H
+  : never;
+
+export type ExtractEvenHandler<Event, T> = Event extends { type: T; payload: infer P }
+  ? P extends any[]
+    ? (...payload: P) => void
+    : never
+  : never;
+export type ExtractEventPayload<Event, T> = Event extends { type: T; payload: infer P } ? P : never;
+
+export class ControllerMessenger<
+  Action extends { type: string; handler: (...args: any) => unknown },
```

Comment:
This is the `any` that I don't know how to get rid of :thinking: Changing this to `unknown` or `unknown[]` breaks everything.